labs/idp-okta-integration/readme.md

```diff
@@ -0,0 +1,182 @@
+![Create App](assets/okta-app-integration.png)
+4. In the next screen, configure the following
+     * App integration name: Apigee App
+     * Sign-in redirect URIs: https://{env group hostname}/v1/oauth20/callback 
+          * example: https://34.149.2.239.nip.io/v1/oauth20/callback
+          * This url points to the idp facade that will be deployed to Apigee
+     * Controlled access: Allow everyone...
+![App Name](assets/okta-app-name.png)
+![Redirect](assets/okta-app-redirect.png)
+![Assignments](assets/okta-assignments.png)
+5. Click **Save** to be taken to the properties of the Okta app just created
+6. Copy the Client ID and Client secret, and Okta domain for use later
+![App Properties](assets/okta-app-properties.png)
+
+
+## Identity Facade Setup
+
+1. In a command line, clone the [devrel repo](https://github.com/apigee/devrel)
+
+```
+git clone https://github.com/apigee/devrel.git
+```
+
+2. Navigate to the identity facade directory
+     
+```
+cd devrel/references/identity-facade
+```
+3. Run the following commands to set the environment variables required for the setup
+
+```export IDP_DISCOVERY_DOCUMENT=https://dev-63386416.okta.com/.well-known/openid-configuration
+export TEST_IDP_APIGEE_CLIENT_ID=0oa2hu6lb5HPR37fY5d7
+export TEST_IDP_APIGEE_CLIENT_SECRET=veERjgCwQ9KUGLidVevKjAwywxX2JBhmg1iQ88gd
+export APIGEE_X_ORG=teodlh-apigeex
+export APIGEE_X_ENV=eval
+export APIGEE_X_HOSTNAME=34.149.2.239.nip.io
+```
+4. Verify the variables have been set correctly
+
+```
+echo $IDP_DISCOVERY_DOCUMENT
+echo $TEST_IDP_APIGEE_CLIENT_ID
+echo $TEST_IDP_APIGEE_CLIENT_SECRET
+echo $APIGEE_X_ORG
+echo $APIGEE_X_ENV
+echo $APIGEE_X_HOSTNAME
+````
+
+5. Deploy the identity facade
+
+```./pipeline.sh --googleapi```
+
+6. The output is show below. Copy and save the following values:
+
+     * consumerKey
+     * consumerSecret
+     * authorization URL
+
+     ![idp facade output](assets/idp-facade-output.png)
+7. Finally, generated environment variables for use later.
+
+     * Base64 encoding of client id and secret for use during Basic Auth in tests
+```
+export APIGEE_CLIENT_ID=xkey-1636042329
+export APIGEE_SECRET=xsecret
+export BASE64_ENCODED=$(echo -n $APIGEE_CLIENT_ID:$APIGEE_SECRET | base64)
+```
+
+
+## Test Identity Facade
+
+This test will simulate a three-legged [OAuth 2.0](https://cloud.google.com/apigee/docs/api-platform/security/oauth/oauth-introduction) flow / authorization grant
+
+1. Open an incognito tab in a browser and visit the authorization URL captured above
+2. Apigee will redirect to Okta to generate an authorization code. Log in using the Okta credentials for the user created earlier.
+     
+     * In a real-world scenario, the client application would build the authorization URL and invoke the flow
+![Okta Auth](assets/okta-auth-code-login.png)
+3. The authorization URL has Okta redirect to httpbin.org which is a 3rd party service that will show payload from Okta. Specifically, we need the authorization code.
```

Comment:
Okta does not redirect to https://httpbin.org/get
Okta redirects to the Apigee callback URL (/v1/oauth20/callback), which controls the incoming query parameters, generate an authorization code (using the same value as the one provided by Okta) and performs a redirection on the client app redirect_uri (here https://httpbin.org/get) providing the authorization_code and initial state parameters